src/Features/LanguageServer/Protocol/Handler/InlineCompletions/XmlSnippetParser.cs

```diff
@@ -0,0 +1,120 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Composition;
+using System.IO;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Snippets;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.InlineCompletions;
+
+/// <summary>
+/// Server instance agnostic snippet parser and cache.
+/// This can be re-used across LSP servers as we're just storing an
+/// internal representation of an XML snippet.
+/// </summary>
+[Export(typeof(XmlSnippetParser)), Shared]
+internal partial class XmlSnippetParser
+{
+    /// <summary>
+    /// Cache to hold onto the parsed XML for a particular snippet.
+    /// </summary>
+    private readonly ConcurrentDictionary<string, ParsedXmlSnippet?> _parsedSnippetsCache = new();
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public XmlSnippetParser()
+    {
+    }
+
+    internal ParsedXmlSnippet? GetParsedXmlSnippet(SnippetInfo matchingSnippetInfo, RequestContext context)
+    {
+        if (_parsedSnippetsCache.TryGetValue(matchingSnippetInfo.Title, out var cachedSnippet))
+        {
+            if (cachedSnippet == null)
+            {
+                context.TraceWarning($"Returning a null cached snippet for {matchingSnippetInfo.Title}");
+            }
+
+            return cachedSnippet;
+        }
+
+        ParsedXmlSnippet? parsedSnippet = null;
+        try
+        {
+            context.TraceInformation($"Reading snippet for {matchingSnippetInfo.Title} with path {matchingSnippetInfo.Path}");
+            parsedSnippet = GetAndParseSnippetFromFile(matchingSnippetInfo);
+            return parsedSnippet;
+        }
+        catch (Exception ex) when (FatalError.ReportAndCatch(ex, ErrorSeverity.General))
+        {
+            context.TraceError($"Got exception parsing xml snippet {matchingSnippetInfo.Title} from file {matchingSnippetInfo.Path}");
+            context.TraceException(ex);
+            return null;
```

Comment:
Nit: Don't need this return, or the one above, or the finally.. can just return `parsedSnippet` at the end. Would make it a little easier to follow

Reply:
🤦fixed!